### PR TITLE
`testers`: docs and move

### LINKED ISSUE #156785: Level up the test functions
**Project description**

 - [ ] move generic testing functions into a separate file. (which may import other files of course)
 - [ ] document them in the Nixpkgs manual

Some functions, off the top of my head:

 - [x] testVersion
 - [x] testEqualDerivation
 - [x] invalidateFetcherByDrvHash
 - [ ] #156613 
 - [ ] nixosTest
 - [ ] nixosTests

### PR DESCRIPTION
**Pull request by roberth**
###### Description of changes

Closes #156785

Tidy up the testers a bit.

- move `invalidateFetcherByDrvHash` into place as expected from docs
- move `nixosTest` to `testers` too and document it in the manual

image

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: doc/builders/testers.chapter.md [hunk 1]
```diff
@@ -80,3 +80,49 @@ tests.fetchgit = invalidateFetcherByDrvHash fetchgit {
...
+}
+```
+
+### Result
+
+A derivation that runs the VM test.
+
+Notable attributes:
+
+ * `nodes`: the evaluated NixOS configurations. Useful for debugging and exploring the configuration.
```
Comment (ncfavier) [inline doc/builders/testers.chapter.md:126]: Note that in https://github.com/NixOS/nixpkgs/pull/164660 I changed `nodes` from being a function that takes a qemu package to that function applied to `pkgs.qemu_test`. I didn't give it much thought back then, and I don't know what we actually want to expose here.

**Review by ncfavier — COMMENTED** [review]
LGTM, though there are still a couple uses of the top-level `nixosTest` alias left.

File: doc/builders/testers.chapter.md (hunk 1 above)
Comment (roberth) [reply doc/builders/testers.chapter.md:126]: I guess that change didn't invalidate the spirit of this doc. Did you have a specific reason to change it, or was it for implementation simplicity?

I think we could change it back when making the tests more flexible as I've proposed in #171396. The qemu package will be configurable, so if you want to apply it to another qemu, you can do so by defining the option value instead. I'm leaning towards leaving this as is for the time being.

Comment (ncfavier) [reply doc/builders/testers.chapter.md:126]: > Did you have a specific reason to change it, or was it for implementation simplicity?

It was just for simplicity.

**Review by Artturin — APPROVED** [review]

**Pull request merged**